rasa_core/featurizers.py

```diff
@@ -163,18 +136,425 @@ def encode(self, active_features, input_feature_map):
+        return X, true_lengths
+
+    def persist(self, path):
+        featurizer_file = os.path.join(path, "featurizer.json")
+        utils.create_dir_for_file(featurizer_file)
+        with io.open(featurizer_file, 'w') as f:
+            f.write(str(jsonpickle.encode(self)))
+
+    @staticmethod
+    def load(path):
+        featurizer_file = os.path.join(path, "featurizer.json")
+        if os.path.isfile(featurizer_file):
+            with io.open(featurizer_file, 'r') as f:
+                _json = f.read()
+            return jsonpickle.decode(_json)
+        else:
+            logger.info("Couldn't load featurizer for policy. "
+                        "File '{}' doesn't exist. ".format(featurizer_file))
+            return None
+
+
+class FullDialogueFeaturizer(Featurizer):
+    def __init__(self, featurize_mechanism):
+        # type: (FeaturizeMechanism) -> None
+        super(FullDialogueFeaturizer, self).__init__(featurize_mechanism)
+
+        self.max_len = None
+
+    def _calculate_max_len(self, as_states):
+        self.max_len = 0
+        for states in as_states:
+            self.max_len = max(self.max_len, len(states))
+        logger.info("The longest dialogue has {} actions."
+                    "".format(self.max_len))
+
+    def _pad_states(self, states):
+        # pad up to max_len
+        if len(states) < self.max_len:
+            states += [None] * (self.max_len - len(states))
+
+        return states
+
+    def training_states_and_actions(
+            self,
+            trackers,  # type: List[DialogueStateTracker]
+            domain  # type: Domain
+    ):
+        # type: (...) -> Tuple[List[List[Dict]], List[List[Dict]], Dict]
+
+        trackers_as_states = []
+        trackers_as_actions = []
+        events_metadata = defaultdict(set)
+
+        for tracker in trackers:
+            states = domain.states_for_tracker_history(tracker)
+
+            delete_first_state = False
+            actions = []
+            for event in tracker._applied_events():
+                if isinstance(event, ActionExecuted):
+                    if not event.unpredictable:
+                        # only actions which can be predicted at a stories start
+                        actions.append(event.action_name)
+                    else:
+                        # unpredictable actions can be only the first in the story
+                        delete_first_state = True
+                else:
+                    action_name = tracker.latest_action_name
+                    events_metadata[action_name].add(event)
+
+            if delete_first_state:
+                states = states[1:]
+
+            trackers_as_states.append(states[:-1])
+            trackers_as_actions.append(actions)
+
+        if self.max_len is None:
+            self._calculate_max_len(trackers_as_actions)
+
+        # TODO do we need that?
+        metadata = {"events": events_metadata,
+                    "trackers": trackers}
```

Comment:
i think the trackers were needed for the visualisation (but we don't need that anymore if we create the trackers before the featurization)

events metadata is used for warnings when using the bot (e.g. if a slot is set after an action, that has never been set in the training data after this specific action)

Reply:
I would be glad to get rid of that, it makes a lot of functions ugly. If we need it, it is maybe make sense to add such method in tracker

Reply:
I think the generator should return events metadata, and then we can pass it to ensemble for persist